pennylane/ops/functions/assert_valid.py

```diff
@@ -56,6 +56,9 @@ def _check_decomposition(op, skip_wire_mapping):
 
         assert isinstance(decomp, list), "decomposition must be a list"
         assert isinstance(compute_decomp, list), "decomposition must be a list"
+        assert op.__class__ not in [
+            decomp_op.__class__ for decomp_op in decomp
+        ], "an operator should not be included in its own decomposition"
```

Comment:
Could there be iterative decompositions, where a decomp contains an operator of its own type in one iteration, but then, upon further decomposition, it decomposes to other stuff?
I'm thinking something like a decomposition that reduces the number of qubits until we're down to one, and then eventually decomposes to `Rot` or so. I'm not aware of an implemented example, so we can always come back to this assertion and change it once we need that.

Reply:
I guess philosophically yes? But let's just update it when it becomes incorrect.